tools/system_libs.py

```diff
@@ -78,161 +78,299 @@ def create_lib(libname, inputs):
+      'scalbn.c', '__fpclassifyl.c',
+      '__signbitl.c', '__signbitf.c', '__signbit.c'
+    ])
+  string_files = files_in_path(
+    path_components=['system', 'lib', 'libc', 'musl', 'src', 'string'],
+    filenames=['memset.c', 'memmove.c'])
+  other_files = files_in_path(
+    path_components=['system', 'lib', 'libc'],
+    filenames=['emscripten_memcpy.c'])
+  return math_files + string_files + other_files
+
+
+class Library(object):
+  # A name that is not None means a concrete library and not an abstract one.
+  name = None
+  depends = []
+  symbols = set()
+  js_depends = []
+
+  # Build settings
+  emcc = shared.EMCC
+  cflags = ['-Werror']
+  src_dir = None
+  src_files = None
+  src_glob = None
+  src_glob_exclude = None
+  includes = []
+  force_object_files = False
+
+  def in_temp(cls, *args):
+    return os.path.join(shared.get_emscripten_temp_dir(), *args)
+
+  def can_use(self):
+    return True
+
+  def can_build(self):
+    return True
+
+  def get_path(self):
+    return shared.Cache.get(self.get_name(), self.build)
+
+  def get_files(self):
+    if self.src_dir:
+      if self.src_files and self.src_glob:
+        raise Exception('Cannot use src_files and src_glob together')
+
+      if self.src_files:
+        return files_in_path(self.src_dir, self.src_files)
+      elif self.src_glob:
+        return glob_in_path(self.src_dir, self.src_glob, self.src_glob_exclude or ())
+
+    raise NotImplementedError()
+
+  def build_objects(self):
+    commands = []
+    objects = []
+    cflags = self.get_cflags()
+    for src in self.get_files():
+      o = self.in_temp(os.path.basename(src) + '.o')
+      commands.append([shared.PYTHON, self.emcc, src, '-o', o] + cflags)
+      objects.append(o)
+    run_commands(commands)
+    return objects
+
+  def build(self):
+    lib_filename = self.in_temp(self.get_name())
+    create_lib(lib_filename, self.build_objects())
+    return lib_filename
+
+  @classmethod
+  def _inherit_list(cls, attr):
+    # Some properties, like cflags and includes, makes more sense to inherit
+    # via concatenation than replacement.
+    result = []
+    for item in cls.__mro__[::-1]:
+      # Using  __dict__ to avoid inheritance
+      result += item.__dict__.get(attr, [])
+    return result
+
+  def get_cflags(self):
+    cflags = self._inherit_list('cflags')
+    cflags += get_cflags(force_object_files=self.force_object_files)
+
+    if self.includes:
+      cflags += ['-I' + shared.path_from_root(*path) for path in self._inherit_list('includes')]
+
+    return cflags
+
+  def get_base_name(self):
+    return self.name
+
+  def get_ext(self):
+    return 'a' if shared.Settings.WASM_BACKEND and shared.Settings.WASM_OBJECT_FILES else 'bc'
+
+  def get_name(self):
+    return self.get_base_name() + '.' + self.get_ext()
+
+  def get_symbols(self):
+    return self.symbols.copy()
+
+  def get_depends(self):
+    return self.depends
+
+  @classmethod
+  def variations(cls):
```

Comment:
I assume this means variations like a `-mt`, etc.? would be good to add a comment explaining that

Reply:
Renamed this to be more clear.